#Importing Libraries

In [1]:
import os
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings

#Read the dataset

In [2]:
df = pd.read_csv("2015.csv", encoding="ISO-8859-1", on_bad_lines="skip")

In [3]:
#to check number of rows and columns
df.shape

(441456, 330)

In [4]:
pd.set_option('display.max_columns', 500)
df.head()

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE         SEQNO  \
0     1.0     1.0  b'01292015'  b'01'  b'29'  b'2015'    1200.0  2.015000e+09   
1     1.0     1.0  b'01202015'  b'01'  b'20'  b'2015'    1100.0  2.015000e+09   
2     1.0     1.0  b'02012015'  b'02'  b'01'  b'2015'    1200.0  2.015000e+09   
3     1.0     1.0  b'01142015'  b'01'  b'14'  b'2015'    1100.0  2.015000e+09   
4     1.0     1.0  b'01142015'  b'01'  b'14'  b'2015'    1100.0  2.015000e+09   

           _PSU  CTELENUM  PVTRESD1  COLGHOUS  STATERES  CELLFON3  LADULT  \
0  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   
1  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   
2  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   
3  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   
4  2.015000e+09       1.0       1.0       NaN       1.0       2.0     NaN   

   NUMADULT        NUMMEN  NUMWOMEN  CTELNUM1  CELLFON2  CADULT  PVTRESD2  \
0       3.0  1.000000e+00       2.0       NaN       NaN     NaN       NaN   
1       1.0  5.397605e-79       1.0       NaN       NaN     NaN       NaN   
2       2.0  1.000000e+00       1.0       NaN       NaN     NaN       NaN   
3       3.0  1.000000e+00       2.0       NaN       NaN     NaN       NaN   
4       2.0  1.000000e+00       1.0       NaN       NaN     NaN       NaN   

   CCLGHOUS  CSTATE  LANDLINE  HHADULT  GENHLTH  PHYSHLTH  MENTHLTH  POORHLTH  \
0       NaN     NaN       NaN      NaN      5.0      15.0      18.0      10.0   
1       NaN     NaN       NaN      NaN      3.0      88.0      88.0       NaN   
2       NaN     NaN       NaN      NaN      4.0      15.0      88.0      88.0   
3       NaN     NaN       NaN      NaN      5.0      30.0      30.0      30.0   
4       NaN     NaN       NaN      NaN      5.0      20.0      88.0      30.0   

   HLTHPLN1  PERSDOC2  MEDCOST  CHECKUP1  BPHIGH4  BPMEDS  BLOODCHO  CHOLCHK  \
0       1.0       1.0      2.0       1.0      1.0     1.0       1.0      1.0   
1       2.0       1.0      1.0       4.0      3.0     NaN       1.0      4.0   
2       1.0       2.0      2.0       1.0      3.0     NaN       1.0      1.0   
3       1.0       2.0      1.0       1.0      1.0     1.0       1.0      1.0   
4       1.0       1.0      2.0       1.0      3.0     NaN       1.0      1.0   

   TOLDHI2  CVDINFR4  CVDCRHD4  CVDSTRK3  ASTHMA3  ASTHNOW  CHCSCNCR  \
0      1.0       2.0       2.0       2.0      1.0      1.0       2.0   
1      2.0       2.0       2.0       2.0      2.0      NaN       2.0   
2      1.0       7.0       2.0       1.0      2.0      NaN       2.0   
3      1.0       2.0       2.0       2.0      2.0      NaN       2.0   
4      2.0       2.0       2.0       2.0      2.0      NaN       2.0   

   CHCOCNCR  CHCCOPD1  HAVARTH3  ADDEPEV2  CHCKIDNY  DIABETE3  DIABAGE2  SEX  \
0       2.0       1.0       1.0       1.0       2.0       3.0       NaN  2.0   
1       2.0       2.0       2.0       2.0       2.0       3.0       NaN  2.0   
2       1.0       2.0       1.0       2.0       2.0       3.0       NaN  2.0   
3       1.0       2.0       1.0       1.0       2.0       3.0       NaN  2.0   
4       2.0       2.0       1.0       2.0       2.0       3.0       NaN  2.0   

   MARITAL  EDUCA  RENTHOM1  NUMHHOL2  NUMPHON2  CPDEMO1  VETERAN3  EMPLOY1  \
0      1.0    4.0       1.0       2.0       NaN      1.0       2.0      8.0   
1      2.0    6.0       1.0       2.0       NaN      2.0       2.0      3.0   
2      2.0    4.0       1.0       2.0       NaN      1.0       2.0      7.0   
3      1.0    4.0       1.0       2.0       NaN      1.0       2.0      8.0   
4      1.0    5.0       1.0       2.0       NaN      2.0       2.0      8.0   

   CHILDREN  INCOME2  INTERNET  WEIGHT2  HEIGHT3  PREGNANT  QLACTLM2  \
0      88.0      3.0       2.0    280.0    510.0       NaN       1.0   
1      88.0      1.0       1.0    165.0    508.0       NaN       1.0   
2      88.0     9

At this point we have 4,90,772 records and 330 columns. Each record contains an individual's BRFSS survey responses.


In [5]:
#Let's select specific columns only of dataset
df_selected = df[['DIABETE3', '_RFHYPE5', 'TOLDHI2', '_CHOLCHK', '_BMI5', 'SMOKE100', 'CVDSTRK3', '_MICHD',
                  '_TOTINDA', '_FRTLT1', '_VEGLT1', '_RFDRHV5', 'HLTHPLN1', 'MEDCOST', 'GENHLTH', 'MENTHLTH',
                  'PHYSHLTH', 'DIFFWALK', 'SEX', '_AGEG5YR', 'EDUCA', 'INCOME2']]

In [6]:
df_selected.shape

(441456, 22)

In [7]:
df_selected.head()

DIABETE3  _RFHYPE5  TOLDHI2  _CHOLCHK   _BMI5  SMOKE100  CVDSTRK3  _MICHD  \
0       3.0       2.0      1.0       1.0  4018.0       1.0       2.0     2.0   
1       3.0       1.0      2.0       2.0  2509.0       1.0       2.0     2.0   
2       3.0       1.0      1.0       1.0  2204.0       NaN       1.0     NaN   
3       3.0       2.0      1.0       1.0  2819.0       2.0       2.0     2.0   
4       3.0       1.0      2.0       1.0  2437.0       2.0       2.0     2.0   

   _TOTINDA  _FRTLT1  _VEGLT1  _RFDRHV5  HLTHPLN1  MEDCOST  GENHLTH  MENTHLTH  \
0       2.0      2.0      1.0       1.0       1.0      2.0      5.0      18.0   
1       1.0      2.0      2.0       1.0       2.0      1.0      3.0      88.0   
2       9.0      9.0      9.0       9.0       1.0      2.0      4.0      88.0   
3       2.0      1.0      2.0       1.0       1.0      1.0      5.0      30.0   
4       2.0      9.0      1.0       1.0       1.0      2.0      5.0      88.0   

   PHYSHLTH  DIFFWALK  SEX  _AGEG5YR  EDUCA  INCOME2  
0      15.0       1.0  2.0       9.0    4.0      3.0  
1      88.0       2.0  2.0       7.0    6.0      1.0  
2      15.0       NaN  2.0      11.0    4.0     99.0  
3      30.0       1.0  2.0       9.0    4.0      8.0  
4      20.0       2.0  2.0       9.0    5.0     77.0

#Clean the data

In [8]:
#finding missing values
df_selected.isnull().sum()

DIABETE3        7
_RFHYPE5        0
TOLDHI2     59154
_CHOLCHK        0
_BMI5       36398
SMOKE100    14255
CVDSTRK3        0
_MICHD       3942
_TOTINDA        0
_FRTLT1         0
_VEGLT1         0
_RFDRHV5        0
HLTHPLN1        0
MEDCOST         1
GENHLTH         2
MENTHLTH        0
PHYSHLTH        1
DIFFWALK    12334
SEX             0
_AGEG5YR        0
EDUCA           0
INCOME2      3301
dtype: int64

In [9]:
# Drop missing values
missing_percentage = (df_selected.isnull().sum() / len(df_selected)) * 100
print(missing_percentage)

DIABETE3     0.001586
_RFHYPE5     0.000000
TOLDHI2     13.399750
_CHOLCHK     0.000000
_BMI5        8.244989
SMOKE100     3.229087
CVDSTRK3     0.000000
_MICHD       0.892954
_TOTINDA     0.000000
_FRTLT1      0.000000
_VEGLT1      0.000000
_RFDRHV5     0.000000
HLTHPLN1     0.000000
MEDCOST      0.000227
GENHLTH      0.000453
MENTHLTH     0.000000
PHYSHLTH     0.000227
DIFFWALK     2.793936
SEX          0.000000
_AGEG5YR     0.000000
EDUCA        0.000000
INCOME2      0.747753
dtype: float64


In [14]:
# !pip install fancyimpute
# from fancyimpute import KNN

In [ ]:
# # Replace missing values using KNN imputation
# knn_imputer = KNN(k=5)  # k is the number of neighbors to consider
# df_imputed = knn_imputer.fit_transform(df_selected)

# # Convert the result back to a DataFrame
# df_selected_imputed = pd.DataFrame(df_imputed, columns=df_selected.columns)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


In [10]:
df_selected = df_selected.dropna()
df_selected.shape

(343606, 22)

# Modifying and cleanning the values to be more suitable to ML algorithms

Reference:
https://www.cdc.gov/brfss/annual_data/2015/pdf/codebook15_llcp.pdf

For DIABETE3:
((Ever told) you have diabetes (If "Yes" and respondent is female, ask "Was this only when you were pregnant?". If Respondent says pre-diabetes or borderline diabetes, use response code 4.))

- 0 is for no diabetes(2) or only during pregnancy(3)

- 1 is for pre-diabetes or borderline diabetes(4)

- 2 is for yes diabetes(1)

- Remove all 7 (dont know) and Remove all 9 (refused)

In [11]:
# frequency of each unique value
frequency = df_selected['DIABETE3'].value_counts()
print(frequency)

DIABETE3
3.0    284625
1.0     49500
4.0      6516
2.0      2591
7.0       333
9.0        41
Name: count, dtype: int64


In [12]:
df_selected['DIABETE3'] = df_selected['DIABETE3'].replace({2 : 0, 3 : 0, 4 : 1, 1 : 2})
df_selected = df_selected[df_selected.DIABETE3 != 7]
df_selected = df_selected[df_selected.DIABETE3 != 9]
df_selected.DIABETE3.unique()

array([0., 2., 1.])

For _RFHYPE5:
(Adults who have been told they have high blood pressure by a doctor, nurse, or other health professional)

- 0 is No high blood pressure(1)

- 1 is high blood pressure(2)

In [13]:
# frequency of each unique value
frequency = df_selected['_RFHYPE5'].value_counts()
print(frequency)

_RFHYPE5
1.0    191461
2.0    151078
9.0       693
Name: count, dtype: int64


In [14]:
df_selected['_RFHYPE5'] = df_selected['_RFHYPE5'].replace({1 : 0, 2 : 1})
df_selected = df_selected[df_selected._RFHYPE5 != 9]
df_selected._RFHYPE5.unique()

array([1., 0.])

For TOLDHI2:
(Have you EVER been told by a doctor, nurse or other health professional that your blood cholesterol is high?)

- 0 is no high blood cholesterol(2)

- 1 is high blood cholesterol(1)

- Remove all 7 (don't know)

- Remove all 9 (refused)



In [15]:
# frequency of each unique value
frequency = df_selected['TOLDHI2'].value_counts()
print(frequency)

TOLDHI2
2.0    195042
1.0    144790
7.0      2464
9.0       243
Name: count, dtype: int64


In [16]:
df_selected['TOLDHI2'] = df_selected['TOLDHI2'].replace({2 : 0})
df_selected = df_selected[df_selected.TOLDHI2 != 7]
df_selected = df_selected[df_selected.TOLDHI2 != 9]
df_selected.TOLDHI2.unique()

array([1., 0.])

For _CHOLCHK:
(Cholesterol check within past five years)

- 0 is for who does not have cholesterol checked in past 5 years(2 or 3)

- 1 is for who Had cholesterol checked in past 5 years(1)

- 7 or 9 or Missing

In [17]:
# frequency of each unique value
frequency = df_selected['_CHOLCHK'].value_counts()
print(frequency)

_CHOLCHK
1.0    323353
2.0     12137
9.0      4342
Name: count, dtype: int64


In [18]:
df_selected['_CHOLCHK'] = df_selected['_CHOLCHK'].replace({2 : 0, 3 : 0})
df_selected = df_selected[df_selected._CHOLCHK != 9]
df_selected._CHOLCHK.unique()

array([1., 0.])

_BMI5:
Body Mass Index (BMI)
(No changes required, these values are BMI * 100)

In [19]:
df_selected['_BMI5'] = df_selected['_BMI5'].div(100).round(0)
df_selected._BMI5.unique()

array([40., 25., 28., 24., 27., 30., 26., 23., 34., 33., 21., 22., 31.,
       38., 20., 19., 32., 46., 41., 37., 36., 29., 35., 18., 54., 45.,
       39., 47., 43., 55., 49., 42., 17., 16., 48., 44., 50., 59., 15.,
       52., 53., 57., 51., 14., 58., 63., 61., 56., 60., 74., 62., 64.,
       13., 66., 73., 65., 68., 85., 71., 84., 67., 70., 82., 79., 92.,
       72., 88., 96., 81., 12., 77., 95., 75., 91., 69., 76., 87., 89.,
       83., 98., 86., 80., 90., 78., 97.])

SMOKE100:
(Have you smoked at least 100 cigarettes in your entire life? [Note: 5 packs = 100 cigarettes])

- 0: No (2)

- 1: Yes (1)

- Remove all 7 (don't knows)

- Remove all 9 (refused)

In [20]:
# frequency of each unique value
frequency = df_selected['SMOKE100'].value_counts()
print(frequency)

SMOKE100
2.0    185299
1.0    148230
7.0      1237
9.0       724
Name: count, dtype: int64


In [21]:
df_selected['SMOKE100'] = df_selected['SMOKE100'].replace({2 : 0})
df_selected = df_selected[df_selected.SMOKE100 != 7]
df_selected = df_selected[df_selected.SMOKE100 != 9]
df_selected.SMOKE100.unique()

array([1., 0.])

CVDSTRK3:
((Ever told) you had a stroke)

- 0: No (2)

- 1: Yes (1)

- Remove all 7 (dont knows)

- Remove all 9 (refused)


In [22]:
df_selected['CVDSTRK3'] = df_selected['CVDSTRK3'].replace({2 : 0})
df_selected = df_selected[df_selected.CVDSTRK3 != 7]
df_selected = df_selected[df_selected.CVDSTRK3 != 9]
df_selected.CVDSTRK3.unique()

array([0., 1.])

_MICHD:
(Respondents that have ever reported having coronary heart disease (CHD) or myocardial infarction (MI) )

- 0: Did not report having MI or CHD (2)

- 1: Reported having MI or CHD (1)



In [23]:
df_selected['_MICHD'] = df_selected['_MICHD'].replace({2 : 0})
df_selected._MICHD.unique()

array([0., 1.])

_TOTINDA:
(Adults who reported doing physical activity or exercise during the past 30 days other than their regular job)

- 0: No physical activity or exercise in last 30 days (2)

- 1: Had physical activity or exercise (1)

- Remove all 7 or 9 (don't know/refused)

In [24]:
# frequency of each unique value
frequency = df_selected['_TOTINDA'].value_counts()
print(frequency)

_TOTINDA
1.0    236353
2.0     81051
9.0     15397
Name: count, dtype: int64


In [25]:
df_selected['_TOTINDA'] = df_selected['_TOTINDA'].replace({2 : 0})
df_selected = df_selected[df_selected._TOTINDA != 9]
df_selected._TOTINDA.unique()

array([0., 1.])

_FRTLT1:
(Consume Fruit 1 or more times per day)

- 0: No, Consumed fruit less than one time per day (2)

- 1: Yes, Consumed fruit one or more times per day (1)

- Remove all dont knows and missing 9

In [26]:
# frequency of each unique value
frequency = df_selected['_FRTLT1'].value_counts()
print(frequency)

_FRTLT1
1.0    197357
2.0    112411
9.0      7636
Name: count, dtype: int64


In [27]:
df_selected['_FRTLT1'] = df_selected['_FRTLT1'].replace({2 : 0})
df_selected = df_selected[df_selected._FRTLT1 != 9]
df_selected._FRTLT1.unique()

array([0., 1.])

_VEGLT1:
(Consume Vegetables 1 or more times per day)

- 0: No, Consumed vegetables less than one time per day(2)

- 1: Yes, Consumed vegetables one or more times per day(1)

- Remove all dont knows and missing 9

In [28]:
df_selected['_VEGLT1'] = df_selected['_VEGLT1'].replace({2 : 0})
df_selected = df_selected[df_selected._VEGLT1 != 9]
df_selected._VEGLT1.unique()

array([1., 0.])

_RFDRHV5:
(Heavy drinkers (adult men having more than 14 drinks per week and adult women having more than 7 drinks per week))

- 0: No (1)

- 1: Yes (2)

- Remove all dont knows and missing 9

In [29]:
df_selected['_RFDRHV5'] = df_selected['_RFDRHV5'].replace({1 : 0, 2 : 1})
df_selected = df_selected[df_selected._RFDRHV5 != 9]
df_selected._RFDRHV5.unique()

array([0., 1.])

HLTHPLN1:
(Do you have any kind of health care coverage, including health insurance, prepaid plans such as HMOs, or government plans such as Medicare, or Indian Health Service?)

- 0: No (2)

- 1: Yes (1)

- Remove 7 and 9 for don't know or refused

In [30]:
df_selected['HLTHPLN1'] = df_selected['HLTHPLN1'].replace({2 : 0})
df_selected = df_selected[df_selected.HLTHPLN1 != 7]
df_selected = df_selected[df_selected.HLTHPLN1 != 9]
df_selected.HLTHPLN1.unique()

array([1., 0.])

MEDCOST:
(Was there a time in the past 12 months when you needed to see a doctor but could not because of cost?)

- 0: No (2)

- 1: Yes (1)

- Remove 7 for don/t know and 9 for refused

In [31]:
# frequency of each unique value
frequency = df_selected['MEDCOST'].value_counts()
print(frequency)

MEDCOST
2.0    272739
1.0     24970
7.0       362
9.0        74
Name: count, dtype: int64


In [32]:
df_selected['MEDCOST'] = df_selected['MEDCOST'].replace({2 : 0})
df_selected = df_selected[df_selected.MEDCOST != 7]
df_selected = df_selected[df_selected.MEDCOST != 9]
df_selected.MEDCOST.unique()

array([0., 1.])

GENHLTH:
(Would you say that in general your health is:)

- This is an ordinal variable that we will keep.

- Remove 7 and 9 for don't know and refused

In [33]:
# frequency of each unique value
frequency = df_selected['GENHLTH'].value_counts()
print(frequency)

GENHLTH
2.0    101963
3.0     89461
1.0     51838
4.0     38892
5.0     15040
7.0       309
9.0       206
Name: count, dtype: int64


In [34]:
df_selected = df_selected[df_selected.GENHLTH != 7]
df_selected = df_selected[df_selected.GENHLTH != 9]
df_selected.GENHLTH.unique()

array([5., 3., 2., 4., 1.])

MENTHLTH:
(Now thinking about your mental health, which includes stress, depression, and problems with emotions, for how many days during the past 30 days was your mental health not good?)

- Number of days(1-30), we will keep this as it is.

- 0: None (88)

- Remove 77 and 99 for don't know not sure and refused

In [35]:
df_selected['MENTHLTH'] = df_selected['MENTHLTH'].replace({88 : 0})
df_selected = df_selected[df_selected.MENTHLTH != 77]
df_selected = df_selected[df_selected.MENTHLTH != 99]
df_selected.MENTHLTH.unique()

array([18.,  0., 30.,  3.,  5., 15., 10.,  6., 20.,  2., 25.,  1., 29.,
        4.,  7.,  8., 21., 14., 26.,  9., 16., 28., 11., 12., 24., 17.,
       13., 23., 27., 19., 22.])

PHYSHLTH:
(Now thinking about your physical health, which includes physical illness and injury, for how many days during the past 30 days was your physical health not good?)

- 1-30: Number of days

- 0: None (88)

- Remove 77 and 99 for don't know not sure and refused


In [36]:
df_selected['PHYSHLTH'] = df_selected['PHYSHLTH'].replace({88 : 0})
df_selected = df_selected[df_selected.PHYSHLTH != 77]
df_selected = df_selected[df_selected.PHYSHLTH != 99]
df_selected.PHYSHLTH.unique()

array([15.,  0., 30.,  2., 14., 28.,  7., 20.,  3., 10.,  1.,  5., 17.,
        4., 19.,  6., 21., 12.,  8., 25., 27., 22., 29., 24.,  9., 16.,
       18., 23., 13., 26., 11.])

DIFFWALK:
(Do you have serious difficulty walking or climbing stairs?)

- 0: No (2)

- 1: Yes (1)

- Remove 7 and 9 for don't know not sure and refused

In [37]:
df_selected['DIFFWALK'] = df_selected['DIFFWALK'].replace({2 : 0})
df_selected = df_selected[df_selected.DIFFWALK != 7]
df_selected = df_selected[df_selected.DIFFWALK != 9]
df_selected.DIFFWALK.unique()

array([1., 0.])

SEX:
(Indicate sex of respondent.)

- 0: Female (2)

- 1: Male (1)


In [38]:
df_selected['SEX'] = df_selected['SEX'].replace({2 : 0})
df_selected.SEX.unique()

array([0., 1.])

_AGEG5YR:
(Fourteen-level age category)

- This is an ordinal variable that we will keep. (groups of age from 18 to 80 or older)

- Remove 14 because it is don't know or missing.

In [39]:
df_selected = df_selected[df_selected._AGEG5YR != 14]
df_selected._AGEG5YR.unique()

array([ 9.,  7., 11., 10., 13.,  8.,  4.,  6.,  2., 12.,  5.,  1.,  3.])

EDUCA:
(What is the highest grade or year of school you completed?)

- This is an ordinal variable that we will keep.(1 being never attended school or kindergarten only up to 6 being college 4 years or more) Scale here is 1-6

- Remove 9 for refused.

In [40]:
df_selected = df_selected[df_selected.EDUCA != 9]
df_selected.EDUCA.unique()

array([4., 6., 3., 5., 2., 1.])

INCOME2:
(Is your annual household income from all sources: (If respondent refuses at any income level, code "Refused."))

- Variable is already ordinal with 1 being less than dollar 10,000 all the way up to 8 being $75,000 or more

- Remove 77 and 99 for don't know and refused

In [41]:
df_selected = df_selected[df_selected.INCOME2 != 77]
df_selected = df_selected[df_selected.INCOME2 != 99]
df_selected.INCOME2.unique()

array([3., 1., 8., 6., 4., 7., 2., 5.])

#Lets check the shape of dataset now

In [42]:
df_selected.shape

(253680, 22)

In [43]:
df_selected.head()

DIABETE3  _RFHYPE5  TOLDHI2  _CHOLCHK  _BMI5  SMOKE100  CVDSTRK3  _MICHD  \
0       0.0       1.0      1.0       1.0   40.0       1.0       0.0     0.0   
1       0.0       0.0      0.0       0.0   25.0       1.0       0.0     0.0   
3       0.0       1.0      1.0       1.0   28.0       0.0       0.0     0.0   
5       0.0       1.0      0.0       1.0   27.0       0.0       0.0     0.0   
6       0.0       1.0      1.0       1.0   24.0       0.0       0.0     0.0   

   _TOTINDA  _FRTLT1  _VEGLT1  _RFDRHV5  HLTHPLN1  MEDCOST  GENHLTH  MENTHLTH  \
0       0.0      0.0      1.0       0.0       1.0      0.0      5.0      18.0   
1       1.0      0.0      0.0       0.0       0.0      1.0      3.0       0.0   
3       0.0      1.0      0.0       0.0       1.0      1.0      5.0      30.0   
5       1.0      1.0      1.0       0.0       1.0      0.0      2.0       0.0   
6       1.0      1.0      1.0       0.0       1.0      0.0      2.0       3.0   

   PHYSHLTH  DIFFWALK  SEX  _AGEG5YR  EDUCA  INCOME2  
0      15.0       1.0  0.0       9.0    4.0      3.0  
1       0.0       0.0  0.0       7.0    6.0      1.0  
3      30.0       1.0  0.0       9.0    4.0      8.0  
5       0.0       0.0  0.0      11.0    3.0      6.0  
6       0.0       0.0  0.0      11.0    5.0      4.0

Check the class Sizes of the diabetes column using group by

In [44]:
df_selected.groupby(['DIABETE3']).size()

DIABETE3
0.0    213703
1.0      4631
2.0     35346
dtype: int64

#Make feature names more readable

In [46]:
# df_selected.columns.tolist()

In [47]:
#Rename the column name
df = df_selected.rename(columns= {'DIABETE3':'Diabetes_012',
 '_RFHYPE5':'HighBP',
 'TOLDHI2':'HighChol',
 '_CHOLCHK':'CholCheck',
 '_BMI5':'BMI',
 'SMOKE100':'Smoker',
 'CVDSTRK3':'Stroke',
 '_MICHD':'HeartDiseaseorAttack',
 '_TOTINDA':'PhysActivity',
 '_FRTLT1':'Fruits',
 '_VEGLT1':'Veggies',
 '_RFDRHV5':'HvyAlcoholConsump',
 'HLTHPLN1':'AnyHealthcare',
 'MEDCOST':'NoDocbcCost',
 'GENHLTH':'GenHlth',
 'MENTHLTH':'MentHlth',
 'PHYSHLTH':'PhysHlth',
 'DIFFWALK':'DiffWalk',
 'SEX':'Sex',
 '_AGEG5YR':'Age',
 'EDUCA':'Education',
 'INCOME2':'Income'})

In [48]:
df.head()

Diabetes_012  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0           0.0     1.0       1.0        1.0  40.0     1.0     0.0   
1           0.0     0.0       0.0        0.0  25.0     1.0     0.0   
3           0.0     1.0       1.0        1.0  28.0     0.0     0.0   
5           0.0     1.0       0.0        1.0  27.0     0.0     0.0   
6           0.0     1.0       1.0        1.0  24.0     0.0     0.0   

   HeartDiseaseorAttack  PhysActivity  Fruits  Veggies  HvyAlcoholConsump  \
0                   0.0           0.0     0.0      1.0                0.0   
1                   0.0           1.0     0.0      0.0                0.0   
3                   0.0           0.0     1.0      0.0                0.0   
5                   0.0           1.0     1.0      1.0                0.0   
6                   0.0           1.0     1.0      1.0                0.0   

   AnyHealthcare  NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex  \
0            1.0          0.0      5.0      18.0      15.0       1.0  0.0   
1            0.0          1.0      3.0       0.0       0.0       0.0  0.0   
3            1.0          1.0      5.0      30.0      30.0       1.0  0.0   
5            1.0          0.0      2.0       0.0       0.0       0.0  0.0   
6            1.0          0.0      2.0       3.0       0.0       0.0  0.0   

    Age  Education  Income  
0   9.0        4.0     3.0  
1   7.0        6.0     1.0  
3   9.0        4.0     8.0  
5  11.0        3.0     6.0  
6  11.0        5.0     4.0

In [49]:
df.shape

(253680, 22)

In [51]:
df.groupby(['Diabetes_012']).size()
# class imbalance!

Diabetes_012
0.0    213703
1.0      4631
2.0     35346
dtype: int64

#Save file to CSV (diabetes_012 is the target variable)

In [52]:
df.to_csv('diabetes_012_health_indicators_BRFSS2015.csv', sep = ",", index = False)

#Binary Classification of diabtes vs no diabetes

0: No diabetes & pre-diabetes

1: Diabetes

In [54]:
df_binary = df
df_binary['Diabetes_012'] = df_binary['Diabetes_012'].replace({1 : 0})
df_binary['Diabetes_012'] = df_binary['Diabetes_012'].replace({2 : 1})

df_binary = df_binary.rename(columns = {'Diabetes_012':'Diabetes_binary'})
df_binary.Diabetes_binary.unique()

array([0., 1.])

In [55]:
df_binary.head()

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0              0.0     1.0       1.0        1.0  40.0     1.0     0.0   
1              0.0     0.0       0.0        0.0  25.0     1.0     0.0   
3              0.0     1.0       1.0        1.0  28.0     0.0     0.0   
5              0.0     1.0       0.0        1.0  27.0     0.0     0.0   
6              0.0     1.0       1.0        1.0  24.0     0.0     0.0   

   HeartDiseaseorAttack  PhysActivity  Fruits  Veggies  HvyAlcoholConsump  \
0                   0.0           0.0     0.0      1.0                0.0   
1                   0.0           1.0     0.0      0.0                0.0   
3                   0.0           0.0     1.0      0.0                0.0   
5                   0.0           1.0     1.0      1.0                0.0   
6                   0.0           1.0     1.0      1.0                0.0   

   AnyHealthcare  NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex  \
0            1.0          0.0      5.0      18.0      15.0       1.0  0.0   
1            0.0          1.0      3.0       0.0       0.0       0.0  0.0   
3            1.0          1.0      5.0      30.0      30.0       1.0  0.0   
5            1.0          0.0      2.0       0.0       0.0       0.0  0.0   
6            1.0          0.0      2.0       3.0       0.0       0.0  0.0   

    Age  Education  Income  
0   9.0        4.0     3.0  
1   7.0        6.0     1.0  
3   9.0        4.0     8.0  
5  11.0        3.0     6.0  
6  11.0        5.0     4.0

In [70]:
df_binary.shape

(253680, 22)

In [56]:
df_binary.groupby(['Diabetes_binary']).size()
#Class imbalance

Diabetes_binary
0.0    218334
1.0     35346
dtype: int64

In [62]:
#Separate the 0(No Diabetes/Pre-diabetes) and 1(Diabetes)
df_5050_0 = df_binary[df_binary['Diabetes_binary'] == 0]
df_5050_1 = df_binary[df_binary['Diabetes_binary'] == 1]

In [63]:
df_5050_0.count()

Diabetes_binary         218334
HighBP                  218334
HighChol                218334
CholCheck               218334
BMI                     218334
Smoker                  218334
Stroke                  218334
HeartDiseaseorAttack    218334
PhysActivity            218334
Fruits                  218334
Veggies                 218334
HvyAlcoholConsump       218334
AnyHealthcare           218334
NoDocbcCost             218334
GenHlth                 218334
MentHlth                218334
PhysHlth                218334
DiffWalk                218334
Sex                     218334
Age                     218334
Education               218334
Income                  218334
dtype: int64

In [64]:
df_5050_1.count()

Diabetes_binary         35346
HighBP                  35346
HighChol                35346
CholCheck               35346
BMI                     35346
Smoker                  35346
Stroke                  35346
HeartDiseaseorAttack    35346
PhysActivity            35346
Fruits                  35346
Veggies                 35346
HvyAlcoholConsump       35346
AnyHealthcare           35346
NoDocbcCost             35346
GenHlth                 35346
MentHlth                35346
PhysHlth                35346
DiffWalk                35346
Sex                     35346
Age                     35346
Education               35346
Income                  35346
dtype: int64

In [66]:
#Select the 35346 random cases from the 0 (non-diabetes group) as we already have 35346 cases from the diabetes risk group
df_5050_0_rand1 = df_5050_0.take(np.random.permutation(len(df_5050_0))[:35346])

#Append the 35346 1s to the 35346 randomly selected 0s
df_5050 = pd.concat([df_5050_0_rand1, df_5050_1], ignore_index=True)

In [67]:
df_5050.head()

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0              0.0     1.0       1.0        1.0  36.0     1.0     0.0   
1              0.0     0.0       1.0        1.0  27.0     0.0     0.0   
2              0.0     1.0       0.0        1.0  30.0     1.0     0.0   
3              0.0     0.0       0.0        1.0  31.0     0.0     0.0   
4              0.0     0.0       1.0        1.0  30.0     1.0     0.0   

   HeartDiseaseorAttack  PhysActivity  Fruits  Veggies  HvyAlcoholConsump  \
0                   0.0           1.0     1.0      1.0                0.0   
1                   0.0           1.0     1.0      1.0                1.0   
2                   0.0           1.0     1.0      1.0                0.0   
3                   0.0           1.0     0.0      1.0                0.0   
4                   0.0           1.0     0.0      1.0                0.0   

   AnyHealthcare  NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex  \
0            1.0          0.0      2.0       0.0       0.0       0.0  1.0   
1            1.0          0.0      1.0       0.0       0.0       0.0  1.0   
2            1.0          0.0      3.0       0.0      12.0       0.0  1.0   
3            1.0          0.0      3.0       0.0       0.0       1.0  1.0   
4            1.0          0.0      4.0       0.0       5.0       0.0  0.0   

    Age  Education  Income  
0   6.0        5.0     8.0  
1   7.0        6.0     8.0  
2  12.0        4.0     6.0  
3  12.0        6.0     6.0  
4  10.0        5.0     6.0

In [68]:
df_5050.tail()

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
70687              1.0     0.0       1.0        1.0  37.0     0.0     0.0   
70688              1.0     0.0       1.0        1.0  29.0     1.0     0.0   
70689              1.0     1.0       1.0        1.0  25.0     0.0     0.0   
70690              1.0     1.0       1.0        1.0  18.0     0.0     0.0   
70691              1.0     1.0       1.0        1.0  25.0     0.0     0.0   

       HeartDiseaseorAttack  PhysActivity  Fruits  Veggies  HvyAlcoholConsump  \
70687                   0.0           0.0     0.0      1.0                0.0   
70688                   1.0           0.0     1.0      1.0                0.0   
70689                   1.0           0.0     1.0      0.0                0.0   
70690                   0.0           0.0     0.0      0.0                0.0   
70691                   1.0           1.0     1.0      0.0                0.0   

       AnyHealthcare  NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex  \
70687            1.0          0.0      4.0       0.0       0.0       0.0  0.0   
70688            1.0          0.0      2.0       0.0       0.0       1.0  1.0   
70689            1.0          0.0      5.0      15.0       0.0       1.0  0.0   
70690            1.0          0.0      4.0       0.0       0.0       1.0  0.0   
70691            1.0          0.0      2.0       0.0       0.0       0.0  0.0   

        Age  Education  Income  
70687   6.0        4.0     1.0  
70688  10.0        3.0     6.0  
70689  13.0        6.0     4.0  
70690  11.0        2.0     4.0  
70691   9.0        6.0     2.0

In [69]:
#Classes are now perfectly balanced
df_5050.groupby(['Diabetes_binary']).size()

Diabetes_binary
0.0    35346
1.0    35346
dtype: int64

#Save binary dataset and 50-50 binary balanced dataset to csv

In [72]:
#Save the 50-50 balanced dataset to csv
df_5050.to_csv('diabetes_binary_5050split_health_indicators_BRFSS2015.csv', sep = ',', index = False)

In [73]:
#Save the original binary dataset to csv
df_binary.to_csv('diabetes_binary_health_indicators_BRFSS2015.csv', sep = ',', index = False)